sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java

```diff
@@ -362,8 +330,77 @@ public boolean shouldPause() {
         protected final Query query;
         @Nullable protected ScoreDoc lastCollected;
 
+        @Nullable Weight weight = null;
+        @Nullable BulkScorer scorer = null;
+        public int pausedDoc = -1;
+
         public InternalCollectContext(Query query) {
             this.query = query;
         }
     }
+
+    public class InternalIndexSearcher  {
+
+        private final CrateSearchContext context;
+        /* In some cases the collector is paused after the last document is emitted.
+           to prevent double emitting, it's important to track that.
+         */
+        private boolean lastDoc = false;
+        private int currentDocBase = 0;
+
+        public InternalIndexSearcher(CrateSearchContext context) {
+            this.context = context;
+        }
+
+        public void search(Query query, Collector results) throws IOException {
+            if (lastDoc) {
+                return;
+            }
+            if (internalCollectContext.weight == null) {
+                internalCollectContext.weight = context.engineSearcher().searcher().createNormalizedWeight(query);
+
+            } else if (internalCollectContext.pausedDoc != -1) {
+                collect(internalCollectContext.pausedDoc);
+                internalCollectContext.pausedDoc = -1;
+            }
+            search(context.searcher().getTopReaderContext().leaves(), internalCollectContext.weight, results);
```

Comment:
context.searcher() returns a ContextIndexSearcher, which can create the normalizedWeitght and engineSearcher() is a Engine.Searcher() which has the readers